sentry.c

```diff
@@ -0,0 +1,124 @@
+#include <sys/mman.h>
+
+#include "include/common.h"
+#include "include/instructions.h"
+#include "include/regs.h"
+
+struct data {
+	int a;
+	int b;
+	char c[256];
+};
+
+void* setup_sentry(void*, void (*f)(uint32_t*, void*));
+void oop_sentry(struct data*, int);
+void simple_sentry(int);
+
+void oop_sentry(struct data* this, int arg) {
+	printf("OOP: %p %d \n", this, arg);
+}
+
+void simple_sentry(int arg) {
+	printf("Simple: %d \n", arg);
+}
+
+void gen_oop(uint32_t* code, void* function) {
+	intptr_t* code_data = (intptr_t*)code;
+	code_data[255] = (intptr_t)function;
+	code_data[254] = (intptr_t)malloc(sizeof(struct data));
+	
+	uint32_t idx = 0;
+	code[idx++] = auipcc(cs2, 1);
+	code[idx++] = clc_128(cs3, cs2, ((-16) + (2 << 20)));
+	code[idx++] = addi(a1, a0, 0);
+	code[idx++] = addi(a2, a1, 0);
+	code[idx++] = addi(a3, a2, 0);
+	code[idx++] = addi(a4, a3, 0);
+	code[idx++] = addi(a5, a4, 0);
+	code[idx++] = addi(a6, a5, 0);
+	code[idx++] = addi(a7, a6, 0);
+	code[idx++] = clc_128(ca0, cs2, ((-32) + (2 << 20)));
+	code[idx++] = cjalr(cnull, cs3);
+
+
+}
+
+void gen_simple(uint32_t* code, void* function) {
+	intptr_t* code_data = (intptr_t*)code;
+	code_data[255] = (intptr_t)function;
+	
+	uint32_t idx = 0;
+	code[idx++] = auipcc(cs2, 1);
+	code[idx++] = clc_128(cs3, cs2, ((-16) + (2 << 20)));
+	code[idx++] = cjalr(cnull, cs3);
+}
+
+void gen_override(uint32_t* code, void* function) {
+	intptr_t* code_data = (intptr_t*)code;
+	code_data[255] = (intptr_t)function;
+	
+	uint32_t idx = 0;
+	code[idx++] = auipcc(cs2, 1);
+	code[idx++] = clc_128(cs3, cs2, ((-16) + (2 << 20)));
+	code[idx++] = addi(a0, cnull, 42);
+	code[idx++] = cjalr(cnull, cs3);
+}
+
+void* setup_sentry(void* function, void (*generate_code)(uint32_t*, void*)) {
+	uint32_t *code =
+		mmap(NULL, 4096, PROT_READ | PROT_WRITE | PROT_EXEC, MAP_ANON | MAP_PRIVATE, -1, 0);
+	
+	generate_code(code, function);
+	
+	return cheri_sealentry(cheri_setflags(code, 1));
+}
+
+void* setup_normal(void* function, void (*generate_code)(uint32_t*, void*)) {
+	uint32_t *code =
+		mmap(NULL, 4096, PROT_READ | PROT_WRITE | PROT_EXEC, MAP_ANON | MAP_PRIVATE, -1, 0);
+	
+	generate_code(code, function);
+	
+	return cheri_setflags(code, 1);
+}
+
+int main() {
+	puts("0: Simple sentry");
+	puts("1: OOP sentry");
+	puts("2: Override sentry");
+	puts("3: Normal function");
+	puts("4: Fail");
+	printf("Mode: ");
+	uint32_t mode = 0;
+	if(scanf("%u", &mode) == 0) {
+		error("Invalid input");
+	}
+
+	void (*fpointer)(int);
+	switch(mode) {
+		case 0:
+			fpointer = setup_sentry(simple_sentry, gen_simple);
+			break;
+		case 1:
+			fpointer = setup_sentry(oop_sentry, gen_oop);
+			break;
+		case 2:
+			fpointer = setup_sentry(simple_sentry, gen_override);
+			break;
+		case 3:
+			fpointer = setup_normal(simple_sentry, gen_simple);
+			break;
+		case 4: // This case should fail
+			fpointer = setup_sentry(simple_sentry, gen_simple);
+			fpointer = cheri_incoffset(fpointer, 16);
+			break;
+	}
```

Comment:
Adding the following case would document that you have considered the fall-through case and decided against giving it special behaviour. 
```
 default :  // do nothing;
              break; 
```

Reply:
I agree that a `default` case here might be a good idea, but I think the right thing to do would be to `err` and exit immediately if it happens?

Reply:
> I think the right thing to do would be to `err` and exit

Given the call to `fpointer` below that is almost certainly the case.

An alternative hazard removal strategy is to restrict the scope of `fpointer` to each case. This has the added benefit of making data-flow a analysis easie. There would be nothing left to make LLVM ponder the actual state of `fpointer` past the switch statement. This is especially important given the calls to cheri intrinsics (that can have effects on `fpointer`) which make the eventual state of `fpointer` opaque. 

There are also potential failure from `mmap` to account for that may influence the author's course of action here.

Reply:
I have decided to make the default case error and exit:
https://github.com/capablevms/cheri-examples/blob/ecf35f220f6e91e3a187ced602ab67e9da616ea0/sentry.c#L153-L155

Reply:
Isn't that the same as `err(1, "Invalid...")` (I think exit codes are conventionally `>0`)?

Reply:
I didn't know about the `err` function. I am using it now.

Reply:
You should use `EXIT_FAILURE` from `<stdlib.h>`. Generally a return code  `> 0` indicates failure but that is not portable. ( /*disapprovingly stares* at VMS )